File: backend/app/models/education.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text, text
from sqlalchemy import Enum as SAEnum
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base
from app.db.enums import ContentStatus, ContentType, enum_values
from app.db.mixins import TimestampMixin, UUIDMixin


class EducationContent(Base, UUIDMixin, TimestampMixin):
    """Doctor-reviewed educational content assigned to patients.

    Per clinical compliance: every published piece stores the reviewing doctor's
    NMC registration for regulatory audit. ai_disclosure=True requires a UI badge.
    """

    __tablename__ = "kc_education_content"

    title: Mapped[str] = mapped_column(String(255), nullable=False)
    slug: Mapped[str] = mapped_column(String(255), nullable=False, unique=True, index=True)
    content_type: Mapped[ContentType] = mapped_column(
        SAEnum(ContentType, name="content_type", create_type=False, values_callable=enum_values),
        nullable=False,
    )
    condition_categories: Mapped[list[Any]] = mapped_column(
        JSONB, nullable=False, server_default=text("'[]'::jsonb")
    )
    content_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    body_md: Mapped[str | None] = mapped_column(Text, nullable=True)
    reviewed_by_doctor_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("dr_doctors.id", ondelete="SET NULL"),
        nullable=True,
    )
    reviewed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    status: Mapped[ContentStatus] = mapped_column(
        SAEnum(ContentStatus, name="content_status", create_type=False, values_callable=enum_values),
        nullable=False,
        server_default=text("'draft'"),
        index=True,
    )
    ai_disclosure: Mapped[bool] = mapped_column(
        Boolean, nullable=False, server_default=text("false")
    )


class EducationAssignment(Base, UUIDMixin, TimestampMixin):
    """A specific piece of content assigned to a patient by a doctor."""

    __tablename__ = "kc_education_assignments"

    content_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("kc_education_content.id", ondelete="CASCADE"),
        nullable=False,
    )
    patient_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("kc_patients.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    assigned_by_doctor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("dr_doctors.id", ondelete="RESTRICT"),
        nullable=False,
    )
    consultation_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("kc_consultations.id", ondelete="SET NULL"),
        nullable=True,
    )
    read_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    notes: Mapped[str | None] = mapped_column(String(500), nullable=True)

File: mobile/lib/api/education.ts
import { apiFetch } from './client';

export interface EducationContent {
  id: string;
  title: string;
  slug: string;
  content_type: 'article' | 'video' | 'pdf';
  condition_categories: string[];
  content_url: string | null;
  body_md: string | null;
  ai_disclosure: boolean;
  reviewed_at: string | null;
}

export interface EducationAssignment {
  id: string;
  content_id: string;
  consultation_id: string | null;
  notes: string | null;
  read_at: string | null;
  created_at: string;
  content: EducationContent;
}

export interface PatientEducationResponse {
  assignments: EducationAssignment[];
  library: EducationContent[];
  library_total: number;
}

export interface ReadResponse {
  assignment_id: string;
  read_at: string;
}

export async function listPatientEducation(page = 1): Promise<PatientEducationResponse> {
  return apiFetch<PatientEducationResponse>(
    `/v1/clinic/patient/education?page=${page}&page_size=20`,
  );
}

export async function getEducationContent(contentId: string): Promise<EducationContent> {
  return apiFetch<EducationContent>(`/v1/clinic/patient/education/${contentId}`);
}

export async function markAssignmentRead(assignmentId: string): Promise<ReadResponse> {
  return apiFetch<ReadResponse>(`/v1/clinic/patient/education/${assignmentId}/read`, {
    method: 'POST',
  });
}
